File: python/1105.filling-bookcase-shelves.py


class Solution:
    def minHeightShelves(self, books: list[list[int]], shelfWidth: int) -> int:
        INF = int(1e9)
        n = len(books)

        dp = [INF] * (n + 1)
        dp[n] = 0

        for i in range(n - 1, -1, -1):
            w, h = 0, 0
            for j in range(i, n):
                w += books[j][0]
                h = max(h, books[j][1])
                if w > shelfWidth:
                    break
                dp[i] = min(dp[i], h + dp[j + 1])

        return dp[0]

File: cpp/1105.filling-bookcase-shelves.cpp
#include <vector>

using std::vector;

class Solution {
public:
  int minHeightShelves(vector<vector<int>> &books, int shelfWidth) {
    int n = books.size(), INF = int(1e9);
    vector<int> dp(n + 1, INF);
    dp[n] = 0;

    for (int i = n - 1; i >= 0; i--) {
      int w = 0, h = 0;
      for (int j = i; j < n; j++) {
        w += books[j][0];
        h = std::max(h, books[j][1]);
        if (w > shelfWidth)
          break;
        dp[i] = std::min(dp[i], dp[j + 1] + h);
      }
    }

    return dp[0];
  }
};